File: course_2/module_2/examples/dijkstra.go
package main

import (
	"container/heap"
	"fmt"
	"math"
	"sort"
	"strings"
)

// PriorityQueue implements a priority queue for Dijkstra's algorithm.
type PriorityQueue []*Item

// Item represents a node and its priority in the priority queue.
type Item struct {
	Node     string // Node label.
	Priority int    // Priority of the node, used for shortest path computation.
	Index    int    // Index of the item in the priority queue.
}

// Len returns the length of the priority queue.
func (pq PriorityQueue) Len() int { return len(pq) }

// Less compares two items in the priority queue based on their priority.
func (pq PriorityQueue) Less(i, j int) bool { return pq[i].Priority < pq[j].Priority }

// Swap swaps two items in the priority queue.
func (pq PriorityQueue) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
	pq[i].Index = i
	pq[j].Index = j
}

// Push adds an item to the priority queue.
func (pq *PriorityQueue) Push(x any) {
	item := x.(*Item)
	item.Index = len(*pq)
	*pq = append(*pq, item)
}

// Pop removes and returns the item with the highest priority (lowest value).
func (pq *PriorityQueue) Pop() any {
	old := *pq
	n := len(old)
	item := old[n-1]
	item.Index = -1 // Mark as removed.
	*pq = old[0 : n-1]
	return item
}

// Graph represents a weighted graph.
type Graph struct {
	Edges map[string]map[string]int // Adjacency list to store edges and their weights.
}

// NewGraph creates and initializes a new Graph.
func NewGraph() *Graph {
	return &Graph{
		Edges: make(map[string]map[string]int),
	}
}

// AddNode adds a node to the graph without edges.
func (g *Graph) AddNode(node string) {
	if _, exists := g.Edges[node]; !exists {
		g.Edges[node] = make(map[string]int) // Initialize an empty adjacency list.
	}
}

// AddEdge adds a directed edge to the graph.
func (g *Graph) AddEdge(from, to string, weight int) {
	g.AddNode(from)
	g.AddNode(to)
	g.Edges[from][to] = weight // Add the edge with its weight.
}

// Dijkstra computes shortest paths from the start node to all other nodes.
func (g *Graph) Dijkstra(start string) (map[string]int, map[string]string) {
	// Initialize distances map with maximum integer values (infinity).
	distances := make(map[string]int)
	previous := make(map[string]string) // To reconstruct paths.

	for node := range g.Edges {
		distances[node] = math.MaxInt
	}
	distances[start] = 0 // Distance to the start node is 0.

	// Initialize the priority queue and add the start node.
	pq := &PriorityQueue{}
	heap.Push(pq, &Item{Node: start, Priority: 0})

	// Process nodes in the priority queue.
	for pq.Len() > 0 {
		// Extract the node with the smallest distance.
		current := heap.Pop(pq).(*Item)

		// Update distances to neighboring nodes.
		for neighbor, weight := range g.Edges[current.Node] {
			newDistance := distances[current.Node] + weight
			if newDistance < distances[neighbor] { // If a shorter path is found, update it.
				distances[neighbor] = newDistance
				previous[neighbor] = current.Node
				heap.Push(pq, &Item{Node: neighbor, Priority: newDistance}) // Add or update the neighbor in the queue.
			}
		}
	}

	return distances, previous // Return the computed shortest distances and previous nodes for paths.
}

// ReconstructPath reconstructs the shortest path from start to a given target node.
func ReconstructPath(previous map[string]string, start, target string) []string {
	path := []string{}
	for at := target; at != ""; at = previous[at] {
		path = append([]string{at}, path...) // Prepend the node to the path.
		if at == start {
			break
		}
	}
	if len(path) == 0 || path[0] != start {
		return []string{} // Return an empty slice if no path exists.
	}
	return path
}

// joinPath formats a slice of nodes into a string path like "A -> B -> C".
func joinPath(path []string) string {
	return strings.Join(path, " -> ")
}

func main() {
	// Create a new graph.
	graph := NewGraph()

	// Add all nodes explicitly to ensure inclusion of isolated nodes.
	graph.AddNode("A")
	graph.AddNode("B")
	graph.AddNode("C")
	graph.AddNode("D")
	graph.AddNode("E")

	// Add edges to the graph.
	graph.AddEdge("A", "B", 1)
	graph.AddEdge("A", "C", 4)
	graph.AddEdge("B", "C", 2)
	graph.AddEdge("B", "D", 6)
	graph.AddEdge("C", "D", 3)
	graph.AddEdge("C", "E", 5)
	graph.AddEdge("D", "E", 1)

	// Compute shortest distances and paths from node "A".
	distances, previous := graph.Dijkstra("A")

	// Collect all nodes and sort them.
	nodes := make([]string, 0, len(graph.Edges))
	for node := range graph.Edges {
		nodes = append(nodes, node)
	}
	sort.Strings(nodes) // Sort nodes alphabetically.

	// Print the shortest distances from node "A".
	fmt.Println("Shortest distances from A:")
	for _, node := range nodes {
		distance := distances[node]
		if distance == math.MaxInt {
			fmt.Printf("To %s: unreachable\n", node) // Print unreachable nodes.
		} else {
			fmt.Printf("To %s: %d\n", node, distance) // Print the shortest distance.
		}
	}

	// Print the shortest paths from node "A".
	fmt.Println("\nShortest paths from A:")
	for _, node := range nodes {
		path := ReconstructPath(previous, "A", node)
		if len(path) > 0 {
			fmt.Printf("To %s: %s\n", node, joinPath(path))
		} else {
			fmt.Printf("To %s: No path found\n", node)
		}
	}
}

File: course_2/module_2/examples/dijkstra.py
import heapq
import math


class Graph:
    """
    Represents a weighted graph using an adjacency list.
    """

    def __init__(self):
        """Initialize the graph with an empty adjacency list."""
        self.edges = {}

    def add_node(self, node):
        """
        Add a node to the graph.

        Args:
            node: The node to add.
        """
        if node not in self.edges:
            self.edges[node] = {}

    def add_edge(self, from_node, to_node, weight):
        """
        Add a directed edge with a weight to the graph.

        Args:
            from_node: The starting node.
            to_node: The ending node.
            weight: The weight of the edge.
        """
        self.add_node(from_node)
        self.add_node(to_node)
        self.edges[from_node][to_node] = weight


class Dijkstra:
    """
    Implements Dijkstra's algorithm for finding shortest paths in a graph.
    """

    @staticmethod
    def compute_shortest_paths(graph, start_node):
        """
        Compute the shortest paths from the start node to all other nodes.

        Args:
            graph: The graph to process.
            start_node: The starting node.

        Returns:
            distances: A dictionary where the keys are nodes and the values are the shortest distances.
            previous: A dictionary for reconstructing the shortest paths.
        """
        if start_node not in graph.edges:
            raise ValueError(f"Start node '{start_node}' is not in the graph.")

        # Initialize distances to all nodes as infinity.
        distances = {node: math.inf for node in graph.edges}
        distances[start_node] = 0

        # Dictionary to reconstruct paths.
        previous = {node: None for node in graph.edges}

        # Priority queue to store (distance, node) pairs.
        pq = []
        heapq.heappush(pq, (0, start_node))

        while pq:
            # Extract the node with the smallest distance.
            current_distance, current_node = heapq.heappop(pq)

            # If the distance is outdated, skip processing.
            if current_distance > distances[current_node]:
                continue

            # Update distances to neighbors.
            for neighbor, weight in graph.edges.get(current_node, {}).items():
                new_distance = current_distance + weight
                if new_distance < distances[neighbor]:
                    distances[neighbor] = new_distance
                    previous[neighbor] = current_node
                    heapq.heappush(pq, (new_distance, neighbor))

        return distances, previous

    @staticmethod
    def reconstruct_path(previous, start_node, end_node):
        """
        Reconstruct the shortest path from the start node to the end node.

        Args:
            previous: The dictionary of predecessors.
            start_node: The starting node.
            end_node: The ending node.

        Returns:
            A list representing the shortest path.
        """
        path = []
        current = end_node
        while current is not None:
            path.insert(0, current)
            current = previous[current]

        if path[0] != start_node:  # If the path does not start with the start_node.
            return []  # No valid path exists.

        return path


if __name__ == "__main__":
    # Create a graph and add edges.
    graph = Graph()
    graph.add_edge("A", "B", 1)
    graph.add_edge("A", "C", 4)
    graph.add_edge("B", "C", 2)
    graph.add_edge("B", "D", 6)
    graph.add_edge("C", "D", 3)
    graph.add_edge("C", "E", 5)
    graph.add_edge("D", "E", 1)

    # Compute shortest distances and paths from node "A".
    dijkstra = Dijkstra()
    distances, previous = dijkstra.compute_shortest_paths(graph, "A")

    # Print the shortest distances from node "A".
    print("Shortest distances from A:")
    for node, distance in distances.items():
        if distance == math.inf:
            print(f"To {node}: unreachable")
        else:
            print(f"To {node}: {distance}")

    # Print the shortest paths from "A" to all other nodes.
    print("\nShortest paths from A:")
    for node in graph.edges:
        path = dijkstra.reconstruct_path(previous, "A", node)
        if path:
            print(f"To {node}: {' -> '.join(path)}")
        else:
            print(f"To {node}: No path found")
